### DISCUSSION TITLE
How to handle the errors message RMG007 and CS8795 in Mapperly. And reduce to code

### DISCUSSION THREAD

**Discussion by qwertybba** [riok/mapperly#581] [Q&A] [answered]
In my project I am trying to implement Mapperly so I do that  
```
    using Riok.Mapperly.Abstractions;
    namespace Application.Profiles;
    [Mapper]
    public partial class MapperlyMapper
    {
        public partial UsersToRoleDto Map(UsersToRole usersToRole);
        public partial UsersToRoleDto Map(UsersToRole usersToRole)
        {
            return new UsersToRoleDto
            {
                User = usersToRole.User.FullName,
                Role = usersToRole.Role.Name
            };
        }
        public partial UsersToRole Map(UsersToRoleDto usersToRoleDto);
        public partial UsersToRole Map(UsersToRoleDto usersToRoleDto)
        {
            return new UsersToRole
            {
                User = new User { FullName = usersToRoleDto.User },
                Role = new Role { Name = usersToRoleDto.Role }
            };
        }
        public partial UserDto Map(User user);
        public partial User Map(UserDto userDto);
    }
```
Now to the error messages if I remove the [Mapper] attribute then I get these error messages: 

> 01-Error CS8795	Partial method 'MapperlyMapper.Map(User)' must have an
> implementation part because it has accessibility modifiers.  
>02-Error CS8795	Partial method 'MapperlyMapper.Map(UserDto)' must have an
> implementation part because it has accessibility modifiers.

on this line of code
```
    public partial UserDto Map(User user);
    public partial User Map(UserDto userDto);
```
But if I added the [Mapper] attribute I get these error messages:

> 01-Error RMG007	Could not map member
> Domain.Entities.UsersToRoleDto.Role of type string to
> Domain.Entities.UsersToRole.Role of type Domain.Entities.Role
> 
> 02-Error RMG007	Could not map member
> Domain.Entities.UsersToRoleDto.User of type string to
> Domain.Entities.UsersToRole.User of type Domain.Entities.User

on this line of code
`    public partial UsersToRole Map(UsersToRoleDto usersToRoleDto);`
Would it be possible to solve both errors CS8795 and RMG007 at the same time?  
Is there a way to reduce to code, such as using ReverseMap() in AutoMapper?

**Comment by TimothyMakkison** [comment, marked as answer]
Hey, thanks for using Mapperly.

It would help if you shared `User`, `UsersToRoleDto` and `Role` etc but I came up with the following.

Mapperly can't know that you want path `User.FullName` of type `User` to map to path `User` of type `string`. the paths dont match. Instead it tries to map both `User` members together. This fails because it doesn't know how to convert from `User` -> `string`. I suggest you use `MapProperty` [docs](https://mapperly.riok.app/docs/configuration/flattening) or define how to convert from `string` -> `User` and `User` -> `string`

```C#
[Mapper]
public partial class Mapper
{
    [MapProperty("User.FullName", "User")]
    [MapProperty("Role.Name", "Role")]
    public partial UsersToRoleDto Map(UsersToRole usersToRole);

    [MapProperty("User", "User.FullName")]
    [MapProperty("Role", "Role.Name")]
    public partial UsersToRole Map(UsersToRoleDto usersToRoleDto);
}
```

> Is there a way to reduce to code, such as using ReverseMap() in AutoMapper?

Unfortunately Mapperly doesn't have a reverse map substitute 😐

**Reply by qwertybba** [reply]
Hello Timothy Makkison,
These are my classes:
```
public partial class User
{
    public User()
    {
        UsersToRoles = new HashSet<UsersToRole>();
    }

    public string Login { get; set; } = null!;
    public string PasswordUser { get; set; } = null!;
    public string? FullName { get; set; }
    public string? Email { get; set; }
    public int Id { get; set; }
    public bool? IsActive { get; set; }

    public virtual ICollection<UsersToRole> UsersToRoles { get; set; }
}
public partial class UserDto
{
    public UserDto()
    {
        UsersToRoles = new HashSet<UsersToRoleDto>();
    }

    public string Login { get; set; } = null!;
    public string PasswordUser { get; set; } = null!;
    public string? FullName { get; set; }
    public string? Email { get; set; }
    public int Id { get; set; }
    public bool? IsActive { get; set; }
    public virtual ICollection<UsersToRoleDto> UsersToRoles { get; set; }
}
public partial class Role
{
    public Role()
    {
        ControlsToRoles = new HashSet<ControlsToRole>();
        UsersToRoles = new HashSet<UsersToRole>();
        PermissionNavigations = new HashSet<PermissionNavigation>();
    }

    public int Id { get; set; }
    public string Name { get; set; } = null!;

    public virtual ICollection<ControlsToRole> ControlsToRoles { get; set; }
    public virtual ICollection<UsersToRole> UsersToRoles { get; set; }
    public virtual ICollection<PermissionNavigation> PermissionNavigations { get; set; }
}
public partial class RoleDto
{
    public RoleDto()
    {
        ControlsToRoles = new HashSet<ControlsToRoleDto>();
        UsersToRoles = new HashSet<UsersToRoleDto>();
    }

    public int Id { get; set; }
    public string Name { get; set; } = null!;

    public virtual ICollection<ControlsToRoleDto> ControlsToRoles { get; set; }
    public virtual ICollection<UsersToRoleDto> UsersToRoles { get; set; }
}
public partial class UsersToRoleDto
{
    public int Id { get; set; }
    public int IdUser { get; set; }
    public int IdRole { get; set; }
    public string User { get; set; }
    public string Role { get; set; }
}
public partial class UsersToRole
{
    public int Id { get; set; }
    public int IdUser { get; set; }
    public int IdRole { get; set; }
    public virtual Role Role { get; set; } = null!;
    public virtual User User { get; set; } = null!;
}

```

**Reply by TimothyMakkison** [reply]
Thanks, did my suggestion work?

I noticed a couple of things looking at your code. `UsersToRole -> UsersToRoleDto` should work fine if you use `MapProperty("User.FullName", "User")`. 
But I'm not sure how you expect to convert from  `UsersToRoleDto -> UsersToRole` unless you just want `User.Fullname = User` with all the other fields left to default. The same applies to `Role`

Is there a reason why your objects are marked `partial`? Mapperly doesn't need the objects to be marked partial.

**Reply by qwertybba** [reply]
Yes, it works.
Regarding your question about (UsersToRole -> UsersToRoleDto) and (UsersToRoleDto -> UsersToRole) So I used to use AutoMapper(ReverseMap()). I will remove it.
As for your question about why the objects are marked as partial, I am using database first and I use EF Core Power Tools to reverse engineer and the code you see it is generated by [EF Core Power Tools](https://github.com/ErikEJ/EFCorePowerTools)

**Reply by qwertybba** [reply]
if you don't mind, I have one last question
In AutoMapper I used to inject it like this 
```
    public static IServiceCollection AddApplicationServices(this IServiceCollection services)
    {
        services.AddAutoMapper(Assembly.GetExecutingAssembly());
        services.AddMapperly(Assembly.GetExecutingAssembly());
        return services;
    }
```
But if I do the same thing with Mapperly I get this error message:

> Error	CS1061	'IServiceCollection' does not contain a definition for 'AddMapperly' and no accessible extension method 'AddMapperly' accepting a first argument of type 'IServiceCollection' could be found (are you missing a using directive or an assembly reference?)

I have already installed
` <PackageReference Include="Riok.Mapperly" Version="2.9.0-next.3" />`

**Reply by TimothyMakkison** [reply]
Mapperly doesn't have an equivalent `AddAutoMapper` or `IMapper`. Mapperly is 100% source generated and doesn't need to interact with `IServiceCollection`. 

Instead you can define a static mapper and directly call it. ie `CarMapper.MapCarToDto(car);` or construct your instance mapper and use it. (You could probably make `MapperlyMapper` 100% static rn)

If you have an instance mapper and want to share state you can register it manually with `builder.Services.AddSingleton<MapperlyMapper>()` or `builder.Services.AddTransient<MapperlyMapper>()` and inject it into your classes/methods.

**Comment by qwertybba** [comment]
Hello Timothy Makkison,
I am trying to following your advice so I came up with this
```
    [MapProperty("User.FullName", "User")]
    [MapProperty("Role.Name", "Role")]
    public static partial UsersToRoleDto MapUsersToRoleToDto(UsersToRole usersToRole);
    //As far as I understand, MapProperty is not needed
    public static partial UsersToRole MapUsersToRoleFromDto(UsersToRoleDto usersToRoleDto);
```
But I get this error message 

> Error RMG007	Could not map member Domain.Entities.UsersToRoleDto.Role of type string to Domain.Entities.UsersToRole.Role of type Domain.Entities.Role
> Error RMG007	Could not map member Domain.Entities.UsersToRoleDto.User of type string to Domain.Entities.UsersToRole.User of type Domain.Entities.User
another issue if I try to use nameof() like this I get this error message
```
    [MapProperty(nameof(UsersToRole.User.FullName), nameof(UsersToRoleDto.User))]
    [MapProperty(nameof(UsersToRole.Role.Name), nameof(UsersToRoleDto.Role))]
public static partial UsersToRoleDto MapUsersToRoleToDto(UsersToRole usersToRole);
```
I get this error message:

> Error RMG006	Specified member FullName on source type Domain.Entities.UsersToRole was not found
> Error RMG006	Specified member Name on source type Domain.Entities.UsersToRole was not found

**Thank you for your time. Please accept my apologies.**

**Reply by TimothyMakkison** [reply]
> I am trying to following your advice so I came up with this
Error RMG007 Could not map member Domain.Entities.UsersToRoleDto.Role of type string to Domain.Entities.UsersToRole.Role of type Domain.Entities.Role
Error RMG007 Could not map member Domain.Entities.UsersToRoleDto.User of type string to Domain.Entities.UsersToRole.User of type Domain.Entities.User

I think `MapProperty` is still needed no? I'm guessing you had a null reference exception because `User` and `Role` wasn't initalized when you tried to set `FullName` and `Name`. I'm not entirely sure how to fix this. Because the properties `User` annd `Role` are not nullable Mapperly assumes that the constructor has initialized `User` and `Role` members so it is safe to set the properties `Name` and `FullName`. 

Marking Role and User as nullable will cause mapperly to conditionally instantiate the objects, although I'm not sure if this option is available to you

I guess you could create a user defined mapping for `string -> User` etc. This way mapperly will see members `UsersToRoleDto.User` and `UsersToRole.User` and use your defined `string -> User` as the converter.



```C#
public partial UsersToRole Map(UsersToRoleDto usersToRoleDto);
private static User StringToUser(string src) => new() { FullName = src };
private static Role StringToRole(string src) => new() { Name = src };
```

**Generated code**
```C#
var target = new global::Riok.Mapperly.Sample.UsersToRole()
{
    Role = StringToRole(usersToRoleDto.Role),
    User = StringToUser(usersToRoleDto.User)
};
target.Id = usersToRoleDto.Id;
target.IdUser = usersToRoleDto.IdUser;
target.IdRole = usersToRoleDto.IdRole;
return target;
```

@latonz do you have a better solution? Has the option to check if a field has been instantiated before setting its children ever been suggested?

> another issue if I try to use nameof() like this I get this error message

Unfortunately you can't use `nameof` for the full member path, `nameof(UsersToRole.User.FullName)` returns "FullName". `nameof` always returns the final property name in the expression. This won't work as mapperly requires the full member path to map properties.
Mapperly may add a "magic" `nameof` handler which would let this work! See #518


```C#
[MapProperty("User", "User.FullName")]
[MapProperty("Role", "Role.Name")]
public partial UsersToRole Map(UsersToRoleDto usersToRoleDto);
```

> Thank you for your time. Please accept my apologies.

No worries, hope this helps 😄

**Comment by qwertybba** [comment]
I have marked Role and User as nullable but I am still getting the error.
```
public partial class UsersToRole
{
    public int Id { get; set; }
    public int IdUser { get; set; }
    public int IdRole { get; set; }

    public virtual Role? Role { get; set; }
    public virtual User? User { get; set; }
}
```
On this line of code `public static partial UsersToRoleDto MapUsersToRoleToDto(UsersToRole usersToRole); ` I do not get any error even when not using MapProperty.
On the other hand, on this line of code `public static partial UsersToRole MapUsersToRoleFromDto(UsersToRoleDto usersToRoleDto); ` I get the error RMG007
Is there a way to get around the error without using MapProperties or creating defined mappings since they are not needed
